network/src/noise/handshake.rs

```diff
@@ -245,42 +263,97 @@ impl NoiseUpgrader {
     /// that successfully authenticate to a public key in our `trusted_peers` set.
     /// In addition, we will expect the client to include an anti replay attack
     /// counter in the Noise handshake payload in mutual auth scenarios.
-    pub async fn upgrade_inbound<TSocket>(
+    pub async fn upgrade_inbound_with_noise<TSocket>(
         &self,
         mut socket: TSocket,
-    ) -> io::Result<NoiseStream<TSocket>>
+    ) -> io::Result<(NoiseStream<TSocket>, PeerId)>
     where
         TSocket: AsyncRead + AsyncWrite + Unpin,
     {
+        // receive [remote_peer_id | self_public_key] as prologue
+        let mut prologue = [0; PeerId::LENGTH + x25519::PUBLIC_KEY_SIZE];
+        socket.read_exact(&mut prologue).await?;
+        let (remote_peer_id, self_expected_public_key) = prologue.split_at(PeerId::LENGTH);
```

Comment:
ahh, same thing here. we should probably just read everything at once?

Reply:
less system calls I guess